### PR TITLE
Add an SSH Factory parameter resolver for unsupported SCM providers

### PR DESCRIPTION
**Pull request by vinokurig**
### What does this PR do?
* Return the SSH factory resolver in order to handle SSH urls from unsupported SCM providers.
* Add a `priority` value for all factory resolvers to be able to control resolver's priority.

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/22488

### How to test this PR?

1. Deploy che from the pr image: `quay.io/eclipse/che-server:pr-567`
2. [Configure SSH keys](https://github.com/devfile/devworkspace-operator/blob/main/docs/additional-configuration.adoc#configuring-devworkspaces-to-use-ssh-keys-for-git-operations) for an unsupported SCM provider e.g. gitea.com.
3. Start a workspace from SSH url of the repository.

See: workspace starts from the default devfile.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolver.java [hunk 1]
```diff
@@ -36,55 +35,42 @@
...
-import org.eclipse.che.commons.annotation.Nullable;
 
 /**
  * Provides Factory Parameters resolver for github repositories.
  *
  * @author Florent Benoit
  */
 @Singleton
-public class GithubFactoryParametersResolver extends RawDevfileUrlFactoryParameterResolver {
+public class GithubFactoryParametersResolver implements FactoryParametersResolver {
```
Comment (tolusha) [inline wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolver.java:45]: Why do we need this changes?

Comment (vinokurig) [reply wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolver.java:45]: To not to override the `priority()` function for each SCM resolver. This function is default in the interface and returns `DEFAULT` priority for each resolver.

File: wsmaster/che-core-api-factory-git-ssh/src/main/java/org/eclipse/che/api/factory/server/git/ssh/GitSshFactoryParametersResolver.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+import org.eclipse.che.api.workspace.shared.dto.devfile.ProjectDto;
+import org.eclipse.che.api.workspace.shared.dto.devfile.SourceDto;
+
+/**
+ * Provides Factory Parameters resolver for Git Ssh repositories.
+ *
+ * @author Anatolii Bazko
+ */
+@Singleton
+public class GitSshFactoryParametersResolver extends RawDevfileUrlFactoryParameterResolver {
```
Comment (tolusha) [inline wsmaster/che-core-api-factory-git-ssh/src/main/java/org/eclipse/che/api/factory/server/git/ssh/GitSshFactoryParametersResolver.java:44]: `extends FactoryParametersResolver` ?

File: wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolver.java (hunk 1 above)
Comment (tolusha) [reply wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolver.java:45]: Ok, but why only for` GithubFactoryParametersResolver` ?

Comment (vinokurig) [reply wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubFactoryParametersResolver.java:45]: Ah, I forgot to apply the interface for other resolvers, fixed.

**Comment by artaleks9** [comment]
* Verified the functionality on Che deployed with specific image `quay.io/eclipse/che-server:pr-567`
This works correctly, expected behaviior.
* Looks good to merge.

**Review by artaleks9 — APPROVED** [review]

**Comment by tolusha** [comment]
Please ignore `ci/prow/v12-git-no-pat-oauth-flow`. It has not been added yet.

File: assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java [hunk 3]
```diff
@@ -183,6 +190,7 @@ protected void configure() {
     scmFileResolverResolverMultibinder.addBinding().to(GitlabScmFileResolver.class);
     scmFileResolverResolverMultibinder.addBinding().to(BitbucketServerScmFileResolver.class);
     scmFileResolverResolverMultibinder.addBinding().to(AzureDevOpsScmFileResolver.class);
+    scmFileResolverResolverMultibinder.addBinding().to(GitSshScmFileResolver.class);
```
Comment (tolusha) [inline assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java:193]: Should we add `RawDevfileUrlFactoryParameterResolver` here?

Comment (vinokurig) [reply assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java:193]: `RawDevfileUrlFactoryParameterResolver ` does not implement `ScmFileResolver`

**Review by tolusha — APPROVED** [review]

**Pull request merged**